datalad/support/annexrepo.py

```diff
@@ -1413,95 +1410,107 @@ def add(self, files, git=None, backend=None, options=None, commit=False,
         # Note: As long as we support direct mode, one should not call
         # super().add() directly. Once direct mode is gone, we might remove
         # `git` parameter and call GitRepo's add() instead.
-        if dry_run:
 
-            git_options = ['--dry-run', '-N', '--ignore-missing']
+        def _get_to_be_added_recs(paths):
+
+            if self.is_direct_mode():
+                # we already know we can't use --dry-run
+                return self._process_git_get_output(
+                    linesep.join(["'{}'".format(p) for p in paths]))
+            else:
+                # Note: if a path involves a submodule in direct mode, while we
+                # are not in direct mode at current level, we might still fail.
+                # Hence the except clause is still needed. However, this is
+                # unlikely, since direct mode usually should be used only, if it
+                # was enforced by FS and/or OS and therefore concerns the entire
+                # hierarchy.
+                _git_options = ['--dry-run', '-N', '--ignore-missing']
+                try:
+                    return super(AnnexRepo, self).add(
+                        files, git_options=_git_options, update=update)
+                except CommandError as e:
+                    if "fatal: This operation must be run in a work tree" \
+                       in e.stderr and \
+                       "fatal: 'git status --porcelain' failed in submodule" \
+                       in e.stderr:
+
+                        lgr.warning(
+                            "Known bug in direct mode."
+                            "We can't use --dry-run when there are submodules in "
+                            "direct mode, because the internal call to git status "
+                            "fails. To be resolved by using (Dataset's) status "
+                            "instead of a git-add --dry-run altogether.")
+                        # fake the return for now
+                        return self._process_git_get_output(
+                            linesep.join(["'{}'".format(f) for f in files]))
+                    else:
+                        # unexpected failure
+                        raise e
```

Comment:
this entire exception handling is not even smoke tested according to codecov report

Reply:
Yes. And it was covered before.

The entire point of the RF'ing is to not go there in any "regular" case. Which is why that doesn't happen anymore in the tests. So in fact the coverage report confirms the RF'ing worked. ;-)

Reply:
1. if not used -- should be removed???
2. I still see `git_options=` so some evil soul could even pass --dry-run?

Reply:
1. Well, it still CAN happen. As noted in the comments this will be the case when we are in a non-direct-mode annex but there's a submodule in direct mode, which then is touched by that dry-run call. That would lead to the error. But this means, that some user intentionally switched a part of the hierarchy to direct mode. So, this isn't a "regular" case, where direct mode is enforced either due to a FS considered "crippled" by annex or because it was set to be the default for all annexes.

2. Yes. Why not? If someone thinks he is smarter than us and wants to pass certain options - he shall have the exception he deserves. ;-)
What are you suggesting? To not allow for that option to be passed or to catch the resulting exception the same way (it would affect the call in line 1488)

Reply:
I am just suggesting to have some test which covers this piece of code.  Since it could be triggered (1.) -- I guess you know how ;)  or could actually be done more efficiently with little mocking of GitRepo.add to trigger those exceptions which you expect to be thrown.